File: packages/types/src/sealedbox.test.ts
import { describe, it, expect } from "vitest";
import { generateKeyPair, sealAnonymous, openAnonymous, toHex, fromHex } from "./sealedbox.js";

describe("sealedbox", () => {
  it("round-trips a message", () => {
    const { publicKey, secretKey } = generateKeyPair();
    const message = new TextEncoder().encode(JSON.stringify({ subjectId: "0xabc", principal: "50000" }));

    const sealed = sealAnonymous(message, publicKey);
    const opened = openAnonymous(sealed, publicKey, secretKey);

    expect(new TextDecoder().decode(opened)).toEqual(new TextDecoder().decode(message));
  });

  it("fails to open with the wrong secret key", () => {
    const recipient = generateKeyPair();
    const wrongKey = generateKeyPair();
    const sealed = sealAnonymous(new TextEncoder().encode("secret"), recipient.publicKey);

    expect(() => openAnonymous(sealed, recipient.publicKey, wrongKey.secretKey)).toThrow();
  });

  it("fails to open a tampered ciphertext", () => {
    const { publicKey, secretKey } = generateKeyPair();
    const sealed = sealAnonymous(new TextEncoder().encode("secret"), publicKey);
    sealed[sealed.length - 1] ^= 0xff;

    expect(() => openAnonymous(sealed, publicKey, secretKey)).toThrow();
  });

  it("hex round-trips", () => {
    const bytes = new Uint8Array([0, 1, 2, 254, 255]);
    expect(fromHex(toHex(bytes))).toEqual(bytes);
    expect(toHex(fromHex("0x00ff"))).toEqual("00ff");
  });
});

File: cre/workflow/sealedbox_test.go
package workflow

import (
	"bytes"
	"crypto/rand"
	"testing"

	"golang.org/x/crypto/nacl/box"
)

func TestSealedBox_RoundTrip(t *testing.T) {
	pub, priv, err := box.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("keygen: %v", err)
	}

	message := []byte(`{"subjectId":"0xabc","principal":"50000"}`)

	sealed, err := SealAnonymous(message, pub)
	if err != nil {
		t.Fatalf("seal: %v", err)
	}

	opened, err := OpenAnonymous(sealed, pub, priv)
	if err != nil {
		t.Fatalf("open: %v", err)
	}

	if !bytes.Equal(opened, message) {
		t.Fatalf("round trip mismatch: got %q, want %q", opened, message)
	}
}

func TestSealedBox_WrongRecipientFails(t *testing.T) {
	pubA, _, _ := box.GenerateKey(rand.Reader)
	_, privB, _ := box.GenerateKey(rand.Reader)

	sealed, err := SealAnonymous([]byte("secret"), pubA)
	if err != nil {
		t.Fatalf("seal: %v", err)
	}

	if _, err := OpenAnonymous(sealed, pubA, privB); err == nil {
		t.Fatal("expected opening with the wrong private key to fail")
	}
}

func TestSealedBox_TamperedCiphertextFails(t *testing.T) {
	pub, priv, _ := box.GenerateKey(rand.Reader)

	sealed, err := SealAnonymous([]byte("secret"), pub)
	if err != nil {
		t.Fatalf("seal: %v", err)
	}
	sealed[len(sealed)-1] ^= 0xFF // flip a bit in the authenticated ciphertext

	if _, err := OpenAnonymous(sealed, pub, priv); err == nil {
		t.Fatal("expected tampered ciphertext to fail authentication")
	}
}

func TestSealedBox_TooShortInputFails(t *testing.T) {
	pub, priv, _ := box.GenerateKey(rand.Reader)
	if _, err := OpenAnonymous([]byte("short"), pub, priv); err == nil {
		t.Fatal("expected too-short input to be rejected")
	}
}
